utilities/transactions/transaction_test.cc

```diff
@@ -4030,6 +4030,51 @@ TEST_P(TransactionTest, SavepointTest3) {
   ASSERT_TRUE(s.IsNotFound());
 }
 
+TEST_P(TransactionTest, SavepointTest4) {
+  WriteOptions write_options;
+  ReadOptions read_options;
+  TransactionOptions txn_options;
+  Status s;
+
+  txn_options.lock_timeout = 1;  // 1 ms
+  Transaction* txn1 = db->BeginTransaction(write_options, txn_options);
+  ASSERT_TRUE(txn1);
+
+  txn1->SetSavePoint();  // 1
+  s = txn1->Put("A", "");
+  ASSERT_OK(s);
+
+  txn1->SetSavePoint();  // 2
+  s = txn1->Put("B", "");
+  ASSERT_OK(s);
+
+  s = txn1->PopSavePoint();  // Remove 2
+  ASSERT_OK(s);
+
+  ASSERT_OK(txn1->RollbackToSavePoint());  // Rollback to 1
+
+  // Verify that everything was rolled back.
+  std::string value;
+  s = txn1->Get(read_options, "A", &value);
+  ASSERT_TRUE(s.IsNotFound());
+
+  s = txn1->Get(read_options, "B", &value);
+  ASSERT_TRUE(s.IsNotFound());
+
+  // Nothing should be locked
```

Comment:
Although the keys are not locked, they are still in txn track_keys_. When deleting the transaction, it also unlocks all the tracked keys, which will get us to this line: 
https://github.com/facebook/rocksdb/blob/9625a2bc2b56c92487922c192f4b903083e63c2c/utilities/transactions/transaction_lock_mgr.cc#L620
I am concerned this assert might be invalid if the keys are unlocked via a RollbackToSavePoint.

Reply:
If we determine that the keys were first tracked (ie. locked) in the current savepoint, we actually remove them from `tracked_keys_` so that they don't get unlocked twice:

https://github.com/facebook/rocksdb/blob/9625a2bc2b56c92487922c192f4b903083e63c2c/utilities/transactions/transaction_base.cc#L172

I guess this is kind of interesting for wup if it uses tracked_keys_ now.. It means that on rollback, we won't necessarily have to rollback the entire key set touched, if RollbackToSavepoint has already done the work for us...